### PR TITLE
Archival: Fix non-deterministic history read

### PR DESCRIPTION
**Pull request by yycptt**
https://github.com/uber/cadence/issues/1980

### CODE REVIEW

**Comment by yycptt** [comment]
I think currentPageToken and nextPageToken should be deleted from blob header.

File: service/worker/archiver/history_blob_iterator.go [hunk 1]
```diff
@@ -104,14 +113,15 @@ func (i *historyBlobIterator) Next() (*HistoryBlob, error) {
 	if !i.HasNext() {
 		return nil, errIteratorDepleted
 	}
-	events, nextPersistencePageToken, historyEndReached, err := i.readBlobEvents(i.persistencePageToken)
+	events, nextPersistencePageToken, historyEndReached, numEventsToSkip, err := i.readBlobEvents(i.persistencePageToken, i.numEventsToSkip)
```
Comment (andrewjdawson2016) [inline service/worker/archiver/history_blob_iterator.go]: This method is returning a lot of things now, what do you think about making a struct for the result?

File: service/worker/archiver/history_blob_iterator_test.go [hunk 2]
```diff
@@ -500,6 +500,89 @@ func (s *HistoryBlobIteratorSuite) TestNext_Success_TenCallsToNext() {
 	s.False(itr.HasNext())
 }
 
+func (s *HistoryBlobIteratorSuite) TestReadBlobEvents_Success_SameHistoryDifferentPage() {
```
Comment (andrewjdawson2016) [inline service/worker/archiver/history_blob_iterator_test.go]: Awesome.

File: service/worker/archiver/history_blob_iterator.go (hunk 1 above)
Comment (yycptt) [reply service/worker/archiver/history_blob_iterator.go]: In my draft pr for recording upload progess using heartbeating, the state of the iterator will be a struct and I can make this function return that struct as part of its result. So how about I making the change in that diff?

Comment (andrewjdawson2016) [reply service/worker/archiver/history_blob_iterator.go]: Sounds good

**Comment by andrewjdawson2016** [comment]
You have an interesting thought about deleting currPageToken and nextPageToken on the blob header. Why do you want to do this? My original reason for including these on the blob header was I thought it was a nice property that the keys for current blob and next blob could be constructed given only the blob itself. If we remove these fields then given a blob somewhere in code you cannot reconstruct the key.

**Comment by yycptt** [comment]
Sorry. Ignore what I said. I thought they are persistence page token.

**Pull request merged**